plugins/GSdx/GSTextureCache.cpp

```diff
@@ -2043,3 +2033,119 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
+void GSTextureCache::SetPalette(Source* s, const GSRenderer* renderer, uint16 pal)
+{
+	Palette* p = m_palette_map.GetPalette(m_renderer, pal);
+	s->m_palette = p->GetPaletteGSTexture();
+	s->m_clut = p->GetClut();
+}
+
+// GSTextureCache::Palette
+
+// Default constructor
+GSTextureCache::Palette::Palette() {
+	m_clut = nullptr;
+	m_tex_palette = nullptr;
+}
+
+// Copy constructor
+GSTextureCache::Palette::Palette(const Palette& palette) {
+	m_clut = palette.m_clut;
+	m_tex_palette = palette.m_tex_palette;
+};
+
+// Creates a new palette texture with current clut content, keeping a reference to its copy
+GSTextureCache::Palette::Palette(const GSRenderer* renderer, uint16 pal) {
+	uint16 palette_size = pal * sizeof(uint32);
+	m_clut = (uint32*)_aligned_malloc(palette_size, 64);
+	memcpy(m_clut, (const uint32*)renderer->m_mem.m_clut, palette_size);
+	m_tex_palette = renderer->m_dev->CreateTexture(256, 1);
+	m_tex_palette->Update(GSVector4i(0, 0, pal, 1), m_clut, palette_size);
+}
+
+uint32* GSTextureCache::Palette::GetClut() {
+	return m_clut;
+}
+
+GSTexture* GSTextureCache::Palette::GetPaletteGSTexture() {
+	return m_tex_palette;
+}
+
+// GSTextureCache::PaletteMap
+
+// Default constructor, reverses space in the multimaps
+GSTextureCache::PaletteMap::PaletteMap() {
+	for (auto& multimap : m_multimaps) {
+		multimap.reserve(MAX_SIZE);
+	}
+}
+
+// Hashes the content of the clut
+size_t GSTextureCache::PaletteMap::HashClut(uint16 pal, const uint32* clut) {
+	size_t clut_hash = 3831179159;
+	for (uint16 i = 0; i < pal; i += 4) {
+		clut_hash = (clut_hash + 1488000301) ^ (clut[i] + 33644011);
+		clut_hash = (clut_hash + 3831179159) ^ (clut[i + 1] + 47627467);
+		clut_hash = (clut_hash + 3659574209) ^ (clut[i + 2] + 577038523);
+		clut_hash = (clut_hash + 33644011) ^ (clut[i + 3] + 3491555267);
+	}
+	return clut_hash;
+}
+
+// Retrieves the palette with the desired clut
+GSTextureCache::Palette* GSTextureCache::PaletteMap::GetPalette(const GSRenderer* renderer, uint16 pal) {
+	// Choose which hash map search into:
+	//    pal == 16  : index 0
+	//    pal == 256 : index 1
+	auto& multimap = m_multimaps[pal == 16 ? 0 : 1];
+
+	const uint32* clut = (const uint32*)renderer->m_mem.m_clut;
+
+	// Search for the clut hash into the hash map
+	size_t clut_hash = HashClut(pal, clut);
+
+	// Retrieve all the relevant Palette(s)
+	auto it1 = multimap.equal_range(clut_hash);
+	uint16 palette_size = pal * sizeof(uint32);
+	for (auto it2 = it1.first; it2 != it1.second; ++it2) {
+		// Hash check passed, compare Palette clut content
+		Palette* palette = it2->second;
+		if (GSVector4i::compare64(palette->GetClut(), clut, palette_size)) {
+			// Clut content match, HIT
+			return palette;
+		}
+		// Same clut hash but content mismatch, FALSE HIT
+	}
+
+	// No Palette with matching clut content hash, MISS
+
+	// Create new Palette
+	m_palettes_stack[m_size] = Palette(renderer, pal);
```

Comment:
Is the palette stack really useful ?

Reply:
Well, I was experimenting with different things, I don't really know the performance impact of the stack, but for small objects it may be good (also increase probability of false HIT in the hash map being a cache hit).

Reply:
I understand. I think the cache is already complex enough ;) Could you test the performance without this intermediate structure.